apps/scan/backend/src/scanners/custom/state_machine.ts

```diff
@@ -0,0 +1,1303 @@
+  pollingTimeout: number
+) {
+  return ({ client }: Context) => {
+    assert(client);
+    return timer(0, pollingInterval).pipe(
+      switchMap(async () => {
+        const status = await client.getStatus();
+        debugPaperStatus('Paper status: %o', status);
+        const mapped = status.isOk()
+          ? scannerStatusToEvent(status.ok())
+          : scannerStatusErrorToEvent(status.err());
+        debugPaperStatus('Mapped paper status: %o', mapped);
+        return mapped;
+      }),
+      timeout({
+        each: pollingTimeout,
+        with: () =>
+          throwError(() => new PrecinctScannerError('paper_status_timed_out')),
+      })
+    );
+  };
+}
+
+async function reset({ client }: Context): Promise<void> {
+  assert(client);
+  debug('Resetting hardware');
+  const result = await client.resetHardware();
+  result.unsafeUnwrap();
+}
+
+async function scan({ client, workspace }: Context): Promise<SheetOf<string>> {
+  assert(client);
+  debug('Scanning');
+  const scanResult = await client.scan({
+    wantedScanSide: ScanSide.A_AND_B,
+    resolution: ImageResolution.RESOLUTION_200_DPI,
+    imageColorDepth: ImageColorDepthType.Grey8bpp,
+    formStandingAfterScan: FormStanding.HOLD_TICKET,
+    // The A4 scanner does not support double sheet detection.
+    doubleSheetDetection: DoubleSheetDetectOpt.DetectOff,
+  });
+  debug('Scan result: %o', scanResult);
+  const images = scanResult.unsafeUnwrap();
+
+  /**
+   *
+   */
+  function trimBlackFromTopAndBottomOfImage(
+    image: ImageFromScanner
+  ): ImageFromScanner {
+    const { imageBuffer, imageWidth, imageHeight } = image;
+    const channelCount = imageBuffer.length / (imageWidth * imageHeight);
+
+    let indexOfFirstNonBlackValue = 0;
+
+    for (let i = 0; i < imageBuffer.length; i += 1) {
+      if (imageBuffer[i] !== 0) {
+        indexOfFirstNonBlackValue = i;
+        break;
+      }
+    }
+
+    let indexOfLastNonBlackValue = imageBuffer.length - 1;
+
+    for (let i = imageBuffer.length - 1; i >= 0; i -= 1) {
+      if (imageBuffer[i] !== 0) {
+        indexOfLastNonBlackValue = i;
+        break;
+      }
+    }
+
+    const offsetPerRow = imageWidth * channelCount;
+    const indexOfFirstNonBlackRow = Math.floor(
+      indexOfFirstNonBlackValue / offsetPerRow
+    );
+    const indexOfLastNonBlackRow = Math.floor(
+      indexOfLastNonBlackValue / offsetPerRow
+    );
+    debug(
+      'Trimming black from top and bottom of image (keeping y=%d..%d)',
+      indexOfFirstNonBlackRow,
+      indexOfLastNonBlackRow
+    );
+
+    const trimmed: ImageFromScanner = {
+      ...image,
+      imageBuffer: imageBuffer.slice(
+        indexOfFirstNonBlackRow * offsetPerRow,
+        (indexOfLastNonBlackRow + 1) * offsetPerRow
+      ),
+      imageHeight: image.imageBuffer.length / offsetPerRow,
+    };
+    debug('Trimmed image: %O', trimmed);
+    return trimmed;
+  }
+
+  // FIXME: we should be able to use the image format directly, but the
```

Comment:
should we make a task or something for this?

Reply:
This is here because the Custom scanner does not trim the black borders around the document it scans. `ballot-interpreter-nh-next` can handle trimming it properly, but `ballot-interpreter-nh` and `ballot-interpreter-vx` cannot. There's an argument to be made that maybe `custom-scanner` should handle this and just always return trimmed images. What do you think?